Add hooks for r10k/g10k and ruby -wc.

pillarsdotnet opened the pull request:
This PR allows puppet developers to check the syntax of their r10k Puppetfile as well as the syntax of their ruby code.

Commit:
Add r10k/g10k hooks.

Commit:
Update README.md

Commit:
Add ruby checks.

chriskuehl commented on .pre-commit-hooks.yaml:
```diff
@@ -1,30 +1,61 @@
+
+- id: rubocop
+  additional_dependencies: ['rubocop']
+  description: Check ruby programs for stylistic problems
```
Hmm, I wonder if it would make sense to leave out the two generic (non-Puppet-specific) Ruby hooks (`rubocop` and `ruby-validate`) and instead recommend using Ruby hooks out of a separate hook repository, keeping this repo for hooks specific to Puppet.

The main reason I'm thinking this is that the generic Ruby hook repos are likely going to be able to do a better job at staying up-to-date with the best linters, have them configured in the best way, etc. We can avoid duplicating the work they're already doing, and they're likely to provide a better experience for consumers over the long term.

The nice thing about pre-commit is that it's pretty easy to just add one more hook repo (I see several listed on https://pre-commit.com/hooks.html that contain `rubocop`, for example, though I haven't tried any) by just adding another entry in your project's YAML file. We could even suggest a repo in the README here?

Then again... I guess you could make the same argument about `erb-validate` since my understanding is that `erb` isn't really specific to Puppet. Though we probably wouldn't want to remove that one since there doesn't seem to be any existing replacement at the moment.

What do you think?

pillarsdotnet replied:
I didn't know it was possible to install multiple hook-repos in the same config-file.

pillarsdotnet replied:
Okay; I'll remove the `rubocop` and `ruby-validate` entries, but I think we should reference the preferred versions.

pillarsdotnet replied:
By the way, I tried using the [`ruby-lint`](https://github.com/pre-commit/mirrors-ruby-lint) hook, but it chokes on a [custom fact](https://puppet.com/docs/facter/latest/custom_facts.html) and the [upstream project](https://gitlab.com/yorickpeterse/ruby-lint) appears unmaintained.

pillarsdotnet replied:
I submitted an issue [here](https://gitlab.com/yorickpeterse/ruby-lint/issues/208).

chriskuehl reviewed (approved):
This is awesome, thanks so much for adding these! Just a quick comment I'm curious to hear your opinion on before merging.

pillarsdotnet commented:
Removed the `rubocop` entry; updated README to reference other recommended pre-commit hooks useful for puppet development.

I didn't find another pre-commit hook implementing the `ruby -wc` check, so I left that one in.

chriskuehl commented on ruby-stubs/r10k-validate:
```diff
@@ -0,0 +1,14 @@
+
+status = 0
+ARGV.each do |file|
+  output = `( cd $(dirname "#{file}" && r10k puppetfile check 2>&1)`
```
Oops, just realized -- I think there's a missing end parenthesis after the `dirname "#{file}"`? I get a syntax error when I test out this hook.

pillarsdotnet replied:
Fixed.

Commit:
Don't actually download modules for g10k Puppetfile validation.

chriskuehl reviewed (approved):
Thanks again! Apologies for the slow review on this one.

chriskuehl commented:
Tagged a new version `v2.1.0` with these changes.
